include/seastar/http/internal/content_source.hh

```diff
@@ -52,13 +57,19 @@ public:
         }
         return _inp.read_up_to(_remaining_bytes).then([this] (temporary_buffer<char> tmp_buf) {
             _remaining_bytes -= tmp_buf.size();
+            if (_rep) {
+                _rep->consumed_content += tmp_buf.size();
+            }
```

Comment:
Is it possible for the reply to get the data_source, static_cast it into content_length_source_impl, and check the result? Would be a little more efficient since it's done one time.

Reply:
I would prefer to avoid the `if`. Can this data_source exist after the request is gone? If not we can make it unconditional.

If we can't, then ok.

Reply:
ah, ok, I see. Let me see what can be done there

Reply:
@avikivity , this if shouldn't be there. The httpd (server) request handler also uses the same content-lengh-data-source and it can track the number of unread bytes by the handler as well. Nowadays the server tries to consume the remaining body after the handler is done instead, which is not nice.

Reply:
"Nowadays the server tries to consume the remaining body after the handler is done instead" - this isn't true. We discussed this heavily (for the server case), but have not done this.

Reply:
I meant this:
https://github.com/scylladb/seastar/blob/3699c0ce7ea98dc170e6d05148d900b5a649d882/src/http/httpd.cc#L277-L287
server gets the next portion of data from the budy and if succeeds, then marks the connection for closing.

I guess I used too strong "consume" verb to describe this :(

Reply:
Yes, right. We do not "consume" the rest of the request (which we could have done but didn't) just try to "read" anything more. Note that as I commented above, content_stream() is not the raw socket, it is actually a wrapper which limits itself to the Content-Length or chunked encoding (if used), so this read will return empty() if we're at the real end of the request - we don't hang reading from the socket after the end of the request.